src/webgpu/shader/validation/expression/call/builtin/textureSample.spec.ts

```diff
@@ -0,0 +1,219 @@
+const builtin = 'textureSample';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+
+* test textureSample coords parameter must be correct type
+* test textureSample array_index parameter must be correct type
+* test textureSample coords parameter must be correct type
+* test textureSample offset parameter must be correct type
+* test textureSample offset parameter must be a const-expression
+* test textureSample offset parameter must be between -8 and +7 inclusive
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import {
+  Type,
+  kConcreteIntegerScalarsAndVectors,
+  kConvertableToFloatScalarsAndVectors,
+  isConvertible,
+} from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+type TextureSampleParams = {
+  coords: string;
+  array_index?: boolean;
+  offset?: string;
+};
+
+const kValidTextureSampleParameterTypes: { [n: string]: TextureSampleParams } = {
+  'texture_1d<f32>': { coords: 'f32' },
```

Comment:
This is the second time that Github has thrown a large comment on the floor.
Earlier today I made a suggestion block here that you replaced these string with `f32` -> `Type.f32`, `vec3<f32>` -> `Type.vec3f`, etc. You can then use `elementTypeOf()` to obtain the element type, instead of more string manipulation with `innerType`.

Maybe my other code examples I provided would make more sense had Github not discarded that comment.

Reply:
no worries. thanks for guiding me through these type types.

updated.